torch/csrc/Device.cpp

```diff
@@ -1,235 +1,106 @@
-    }
-    int32_t device_index = -1;
-    if (!r.isNone(1)) {
-      device_index = r.toInt64(1);
-      // -1 is allowed in ATen/C++, to mean the default device, but not in
-      // Python.
-      TORCH_CHECK(device_index >= 0, "Device index must not be negative");
-    }
-    at::Device device(as_device.type(), device_index);
-    return THPDevice_New(device);
-  }
-  Py_RETURN_NONE;
-  END_HANDLE_TH_ERRORS
+inline Device parseDeviceIndex(DeviceType type, DeviceIndex index) {
+  // -1 is allowed in ATen/C++, to mean the default device, but not in Python
+  TORCH_CHECK(index >= 0, "Device index must not be negative");
+  return Device(type, index);
 }
 
-PyObject *THPDevice_type(THPDevice *self, PyObject *noargs)
-{
-  HANDLE_TH_ERRORS
+inline std::string deviceType(const Device& device) {
   std::ostringstream oss;
-  oss << self->device.type();
-  return THPUtils_packString(oss.str().c_str());
-  Py_RETURN_NONE;
-  END_HANDLE_TH_ERRORS
-}
-
-PyObject *THPDevice_index(THPDevice *self, PyObject *noargs)
-{
-  HANDLE_TH_ERRORS
-  if (self->device.has_index()) {
-    return THPUtils_packInt64(self->device.index());
-  } else {
-    Py_RETURN_NONE;
-  }
-  END_HANDLE_TH_ERRORS
-}
-
-static Py_ssize_t THPDevice_hash(THPDevice *self)
-{
-  HANDLE_TH_ERRORS
-  return static_cast<Py_ssize_t>(std::hash<at::Device>{}(self->device) % std::numeric_limits<Py_ssize_t>::max());
-  END_HANDLE_TH_ERRORS_RET(-1)
+  oss << device.type();
+  return oss.str();
 }
 
-PyObject *THPDevice_rc(PyObject *a, PyObject *b, int op) {
-  HANDLE_TH_ERRORS
-  if (!THPDevice_Check(a) || !THPDevice_Check(b)) {
-    // Py_RETURN_NOTIMPLEMENTED not in python 2.
-    Py_INCREF(Py_NotImplemented);
-    return Py_NotImplemented;
-  }
-  THPDevice *da = reinterpret_cast<THPDevice*>(a);
-  THPDevice *db = reinterpret_cast<THPDevice*>(b);
-
-  switch(op) {
-    case Py_EQ:
-      if (da->device == db->device) {
-        Py_RETURN_TRUE;
-      } else {
-        Py_RETURN_FALSE;
-      }
-    case Py_NE:
-      if (da->device == db->device) {
-        Py_RETURN_FALSE;
-      } else {
-        Py_RETURN_TRUE;
-      }
-    case Py_LT:
-    case Py_LE:
-    case Py_GT:
-    case Py_GE:
-      throw torch::TypeError("comparison not implemented");
-    default:
-      throw torch::TypeError("unexpected comparison op");
-  }
-  END_HANDLE_TH_ERRORS
-}
+} // namespace
 
-PyObject *THPDevice_reduce(PyObject *_self, PyObject *noargs)
-{
-  HANDLE_TH_ERRORS
-  auto self = (THPDevice*)_self;
-  auto ret = THPObjectPtr{PyTuple_New(2)};
-  if (!ret) throw python_error();
-
-  py::object torch_module = py::module::import("torch");
-  py::object torch_device = torch_module.attr("device");
-  PyTuple_SET_ITEM(ret.get(), 0, torch_device.release().ptr());
-
-  THPObjectPtr args;
-  std::ostringstream oss;
-  oss << self->device.type();
-  if (self->device.has_index()) {
-    args = THPObjectPtr{Py_BuildValue("(si)", oss.str().c_str(), self->device.index())};
+Device parseDevice(py::object device) {
```

Comment:
The naming here doesn't seem ideal; you're using up prime namespace `torch::parseDevice` for a Python specific functionality (converting python object into device). Maybe you want a pybind11 autocaster here instead?

Reply:
Can I actually do that? I have a feeling that a manual type caster for `at::Device` would clash with the default one pybind11 used for class bindings...

Reply:
Regarding the namespace issue, if we are going to keep `parseDevice` should it live in `torch::util` or `torch::python`? Also will it be better to put `initDeviceBindings` in `torch::python` rather than the top-level `torch` namespace?

Reply:
We have a type caster for Tensor in torch/csrc/utils/pybind.h, so I don't think the conflict will be a problem.

Re namespace, I don't know what the prevailing convention is for Python bindings. A lot of stuff is already dumped in the torch namespace (and I'm trying to make it not worse). One existing convention is to put implementation specific stuff in an impl namespace, e.g., `torch::impl::parseDevice`, though `torch::python` isn't bad either as it indicates that it is Python specific. I'd say that initDeviceBindings is also better moved. torch only for user public APIs.

Reply:
Ok I experimented with a custom type caster for `at::Device` and I don't think it's possible because [pybind11 requires the type of the type caster to be default constructible](https://pybind11.readthedocs.io/en/stable/advanced/cast/custom.html), while `at::Device` don't have a default constructor. We'll have to resort to a separate `parseDevice` function.

As for the namespace, I'll go with putting everything under `torch::python`. `impl` seems vague and doesn't indicate what implementation the code is for...